### PR TITLE
PR #1 - Updated pipeline to add jobs for creating OpenAPI spec file and uploading the artifact. 

### PR DESCRIPTION
**Pull request by pritipath**
Issue Number: (Link to Github Issue or Azure Dev Ops Task/Story)

## Summary
Added support for Swashbuckle to generate API documentation
Updated Annotations in the Controller to accurately document the routes

## Changes

- List of comprehensive changes

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [x] Documentation Updates Made?
- [x] This is not a breaking change. If it is, please describe it below.

## Is this a breaking change?
If yes, what workflow does this break? 

## Anything else?
Other comments, collaborators, etc.
This PR Closes Issue #

### CODE REVIEW

File: src/CarbonAware.WebApi/src/Controllers/CarbonAwareController.cs [hunk 1]
```diff
@@ -58,13 +67,17 @@ public async Task<IActionResult> GetEmissionsDataForLocationsByTime(string locat
...
     }
 
+    /// <summary>
+    /// Calculate the best emission data by location for a specified time period.
+    /// </summary>
+    /// <returns>Array of EmissionsData objects that contains the location, time and the rating in g/kWh</returns>
     [Produces("application/json")]
     [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(IEnumerable<EmissionsData>))]
     [ProducesResponseType(StatusCodes.Status204NoContent)]
     [ProducesResponseType(StatusCodes.Status400BadRequest)]
```
Comment (bderusha) [inline src/CarbonAware.WebApi/src/Controllers/CarbonAwareController.cs]: I think 400 responses should specify `Type = typeof(ValidationProblemDetails)`.  I don't think dotnet will surface that by default

(here and throughout the controlllers"

File: src/CarbonAware/src/Model/EmissionsData.cs [hunk 2]
```diff
@@ -3,9 +3,13 @@
 [Serializable]
 public record EmissionsData
 {
+    ///<example> ueast </example>
```
Comment (bderusha) [inline src/CarbonAware/src/Model/EmissionsData.cs]:
```suggestion
    ///<example> eastus </example>
```

File: src/CarbonAware.WebApi/src/Models/SciScoreInput.cs [hunk 3]
```diff
@@ -18,19 +18,24 @@ public record SciScoreInput
...
 
+    /// <example>153.23</example>
     [JsonPropertyName("latitude")]
     public decimal? Latitude { get; set; }
 
+    /// <example>72.23</example>
     [JsonPropertyName("longitude")]
     public decimal? Longitude { get; set; }
 
+    /// <example>Azure</example>
```
Comment (bderusha) [inline src/CarbonAware.WebApi/src/Models/SciScoreInput.cs:34]: Not blocking, but this value is linked to an enum.  Do you know if/how we can surface enums dynamically in swagger?

**Review by bderusha — CHANGES_REQUESTED** [review]
It wouldn't let me comment on the empty file, but can you explain more about what [src/CarbonAware.WebApi/src/api/v1/test.cs](https://github.com/microsoft/carbon-aware-sdk/pull/73/files#diff-014adba8d85f4dbd54d45be5db8f8083e069f4e3f82b9474c0692b276539ba70) is for?

**Comment by pritipath** [comment]
> It wouldn't let me comment on the empty file, but can you explain more about what [src/CarbonAware.WebApi/src/api/v1/test.cs](https://github.com/microsoft/carbon-aware-sdk/pull/73/files#diff-014adba8d85f4dbd54d45be5db8f8083e069f4e3f82b9474c0692b276539ba70) is for?

Oops, that was checked in by mistake!

File: src/CarbonAware.WebApi/src/Controllers/CarbonAwareController.cs (hunk 1 above)
Comment (pritipath) [reply src/CarbonAware.WebApi/src/Controllers/CarbonAwareController.cs]: I think it does surface by default. Or is this not what you expect - 

image

File: src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            <example>153.23</example>
+        </member>
+        <member name="P:CarbonAware.WebApi.Models.LocationInput.Longitude">
+            <example>72.23</example>
+        </member>
+        <member name="P:CarbonAware.WebApi.Models.LocationInput.CloudProvider">
+            <example>Azure</example>
+        </member>
+        <member name="P:CarbonAware.WebApi.Models.LocationInput.RegionName">
+            <example>useast</example>
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml]: Is this autogenerated?. If not, seems like this should be `eastus` instead

File: src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0"?>
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml]: Is this file needed to be part of the commit?

**Review by juzuluag — COMMENTED** [review]
Few comments. Overall lgtm.

File: .github/workflows/build-webapi.yaml [hunk 6]
```diff
@@ -38,7 +42,15 @@ jobs:
         working-directory: ${{ env.DOTNET_SRC_DIR }}
       - name: Codecov
         uses: codecov/codecov-action@v1
-
+      - name: Generate Open API
+        run: dotnet tool run swagger tofile --output ./api/v1/swagger.yaml --yaml ${{ env.DLL_FILE_PATH }} v1
```
Comment (juzuluag) [inline .github/workflows/build-webapi.yaml:46]: 👍

File: src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml (hunk 4 above)
Comment (pritipath) [reply src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml]: Yes, this file is autogenerated. The Swagger tool uses the xml documentation in this file to generate API spec.

Comment (juzuluag) [reply src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml]: > Yes, this file is autogenerated. The Swagger tool uses the xml documentation in this file to generate API spec.

Does it need to be checked-in as part of the src tree?

File: src/CarbonAware.WebApi/src/api/v1/.gitkeep [hunk 7]
```diff
@@ -0,0 +1 @@
+#dummy file created to commit the new folder. Will remove it in the next PR
```
Comment (JenMadiedo) [inline src/CarbonAware.WebApi/src/api/v1/.gitkeep:1]: What do you expect to go in this folder?

**Review by JenMadiedo — APPROVED** [review]
lgtm

File: src/CarbonAware.WebApi/src/api/v1/.gitkeep (hunk 7 above)
Comment (pritipath) [reply src/CarbonAware.WebApi/src/api/v1/.gitkeep:1]: The generated swagger.yaml file which contains the api spec

File: src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml (hunk 4 above)
Comment (pritipath) [reply src/CarbonAware.WebApi/src/CarbonAware.WebApi.xml]: Removed the file

File: src/CarbonAware.WebApi/src/Models/SciScoreInput.cs (hunk 3 above)
Comment (pritipath) [reply src/CarbonAware.WebApi/src/Models/SciScoreInput.cs:34]: For the example, I think we need to manually specify a value. There are annotations to display all possible enum values if the input parameter is of type enum. In our case, our input is a json object.
I need to do a little more digging on this, but as far the example goes, I think it as to be hardcoded.

**Review by juzuluag — APPROVED** [review]
👍

**Review by bderusha — APPROVED** [review]
Thanks for those changes.  Looks great
